extensions/integration-tests/src/dacpac.test.ts

```diff
@@ -0,0 +1,140 @@
+				await assertFileGenerationResult(packageFilePath);
+
+				assert(extractResult.success === true, 'Extract dacpac should succeed');
+				assert(extractResult.errorMessage === '', `Expected: there should be no error. Actual Error message: "${extractResult.errorMessage}"`);
+			} finally {
+				await utils.deleteDB(databaseName, ownerUri);
+			}
+		});
+
+		test('Import and export bacpac', async function () {
+			let server = await getStandaloneServer();
+			await utils.connectToServer(server);
+
+			let nodes = <azdata.objectexplorer.ObjectExplorerNode[]>await azdata.objectexplorer.getActiveConnectionNodes();
+			assert(nodes.length > 0, `Expecting at least one active connection, actual: ${nodes.length}`);
+
+			let index = nodes.findIndex(node => node.nodePath.includes(server.serverName));
+			assert(index !== -1, `Failed to find server: "${server.serverName}" in OE tree`);
+
+			let ownerUri = await azdata.connection.getUriForConnection(nodes[index].connectionId);
+			let now = new Date();
+			let databaseName = 'ADS_importBacpac_' + now.getTime().toString();
+
+			try {
+				let dacfxService = await azdata.dataprotocol.getProvider<azdata.DacFxServicesProvider>('MSSQL', azdata.DataProviderType.DacFxServicesProvider);
+				assert(dacfxService, 'DacFx Service Provider is not available');
+
+				// Import bacpac
+				let importResult = await dacfxService.importBacpac(path.join(__dirname, 'testData/Database1.bacpac'), databaseName, ownerUri, azdata.TaskExecutionMode.execute);
+				await assertDatabaseCreationResult(databaseName, ownerUri);
+				assert(importResult.success === true, 'Import bacpac should succeed');
+				assert(importResult.errorMessage === '', `Expected: there should be no error. Actual Error message: "${importResult.errorMessage}"`);
+
+				// Export bacpac
+				let folderPath = path.join(process.env.APPDATA, 'DacFxTest');
+				if (!fs.existsSync(folderPath)) {
+					fs.mkdirSync(folderPath);
+				}
+				let packageFilePath = path.join(folderPath, `${databaseName}.bacpac`);
+				let exportResult = await dacfxService.exportBacpac(databaseName, packageFilePath, ownerUri, azdata.TaskExecutionMode.execute);
+				await assertFileGenerationResult(packageFilePath);
+
+				assert(exportResult.success === true, 'Export bacpac should succeed');
+				assert(exportResult.errorMessage === '', `Expected: there should be no error. Actual Error message: "${exportResult.errorMessage}"`);
+			} finally {
+				await utils.deleteDB(databaseName, ownerUri);
+			}
+		});
+	});
+}
+
+async function assertDatabaseCreationResult(databaseName: string, ownerUri: string): Promise<void> {
+	let retryCount = 24; // wait up to two minutes since database can take a long time to get created
+	let result: azdata.SimpleExecuteResult;
+	while (retryCount > 0) {
+		--retryCount;
+		await utils.sleep(5000);
+
+		let query = `BEGIN TRY
+				SELECT name FROM master.dbo.sysdatabases WHERE name='${databaseName}'
```

Comment:
Oh this is because the DacFX operations are async right? There's no event or something else you could hook in to for this? That seems like a useful thing to have if not...

Reply:
yeah the dacfx operations start a task